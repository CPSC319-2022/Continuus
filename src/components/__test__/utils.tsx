import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore, type PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "~/redux/store";
import { reducer } from "~/redux/slices";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
