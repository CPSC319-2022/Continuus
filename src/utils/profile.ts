/**
 * Returns a path to the profile page for the given profile. If the argument is undefined, redirect to the home page.
 *
 * @param profile the user id of the account
 *
 * @return the profile path if the profile is a string, otherwise the home path
 */
export const userPathToProfile = (profile: string | undefined): string => {
    if (profile === undefined) {
        return '/';
    }

    return '/profile/' + encodeURIComponent(profile);
}
