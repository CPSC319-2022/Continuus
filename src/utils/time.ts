export const timeAgo = (inputDate: Date): string => {
  const now = new Date();
  const diffInSeconds = Math.max(Math.floor(
    (now.getTime() - inputDate.getTime()) / 1000
  ), 0);

  if (diffInSeconds < 60) {
    return `${diffInSeconds} ${diffInSeconds === 1 ? "second" : "seconds"} ago`;
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 30) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  const diffInMonths = Math.floor(diffInDays / 30);
  if (diffInMonths < 12) {
    return `${diffInMonths} ${diffInMonths === 1 ? 'month' : 'months'} ago`;
  }

  const diffInYears = Math.floor(diffInMonths / 12);
  return `${diffInYears} ${diffInYears === 1 ? 'year' : 'years'} ago`;
};
