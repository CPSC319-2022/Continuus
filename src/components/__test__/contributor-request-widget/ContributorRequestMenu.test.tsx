import { describe, it, expect, beforeEach, vi } from "vitest";
import { cleanup, render } from "@testing-library/react";
import { renderWithProviders } from "../utils";

vi.mock("next-auth/react", () => ({
  useSession: () => ({
    data: {
      expires: "1",
      user: { id: "test-user-id", email: "a", name: "hi", image: "c" },
    },
  }),
}));

const props = {
  isOpen: true,
  setIsOpen: vi.fn(),
};

const initialMock = {
  api: {
    user: {
      currentUser: {
        useQuery: () => ({
          data: null,
          isLoading: false,
        }),
      },
      updateOne: {
        useMutation: vi.fn(),
      },
    },
    contributorRequest: {
      getAll: {
        useQuery: () => ({
          data: [
            {
              id: "contribreq-id-1",
              createdAt: new Date(),
              user: {
                id: "user-id-1",
                name: "test user 1",
                image: "image",
              },
            },
            {
              id: "contribreq-id-2",
              createdAt: new Date(),
              user: {
                id: "user-id-2",
                name: "test user 2",
                image: "image",
              },
            },
          ],
          isLoading: false,
        }),
      },
      delete: {
        useMutation: () => ({
          mutate: vi.fn(),
          mutateAsync: vi.fn(),
        }),
      },
    },
    useContext: () => ({
      contributorRequest: {
        invalidate: vi.fn(),
      },
    }),
  },
};

describe("Contributor Request Menu: Snapshot", () => {
  beforeEach(() => {
    cleanup();
    vi.clearAllMocks();
    vi.resetModules();
  });

  const setup = (mock: any) => {
    vi.doMock("~/utils/api", () => ({ ...mock }));
  };

  it("render menu open with 2 requests", async () => {
    setup(initialMock);
    const { ContributorRequestMenu } = await import(
      "~/components/contributor-request-widget/ContributorRequestMenu"
    );
    const body = renderWithProviders(<ContributorRequestMenu {...props} />).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render menu open with no requests", async () => {
    setup({
      ...initialMock,
      api: {
        ...initialMock.api,
        contributorRequest: {
          ...initialMock.api.contributorRequest,
          getAll: {
            useQuery: () => ({
              data: [],
              isLoading: false,
            }),
          },
        },
      },
    });
    const { ContributorRequestMenu } = await import(
      "~/components/contributor-request-widget/ContributorRequestMenu"
    );
    const body = renderWithProviders(<ContributorRequestMenu {...props} />).baseElement;
    expect(body).toMatchSnapshot();
  });

  it("render menu closed", async () => {
    setup(initialMock);
    const { ContributorRequestMenu } = await import(
      "~/components/contributor-request-widget/ContributorRequestMenu"
    );
    const body = renderWithProviders(
      <ContributorRequestMenu isOpen={false} setIsOpen={vi.fn} />
    ).baseElement;
    expect(body).toMatchSnapshot();
  });
});
