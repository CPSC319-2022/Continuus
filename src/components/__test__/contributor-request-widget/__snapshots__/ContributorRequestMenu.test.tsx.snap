// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contributor Request Menu: Snapshot > render menu closed 1`] = `
<body>
  <div />
</body>
`;

exports[`Contributor Request Menu: Snapshot > render menu open with 2 requests 1`] = `
<body>
  <div>
    <div
      class="absolute right-0 md:right-5 top-12 w-full max-h-96 md:w-96 rounded-md bg-white p-3 shadow border"
    >
      <div
        class="scrollbar-hide max-h-80 overflow-auto overflow-x-hidden"
      >
        <div
          class="sticky top-0 bg-white"
        >
          <p
            class="text-lg mb-2"
          >
            Contributor Requests
          </p>
          <hr />
        </div>
        <div
          class="py-2 flex justify-between border-b"
        >
          <div
            class="flex"
          >
             
            <a
              href="/profile/user-id-1"
            >
              <img
                alt="avatar"
                class="inline-block aspect-square rounded-[50%] self-center mt-1"
                src="image"
                style="width: 2rem;"
              />
            </a>
            <div
              class="mx-2"
            >
              <p>
                test user 1
              </p>
              <p
                class="text-xs text-gray-400"
              >
                0 seconds ago
              </p>
            </div>
          </div>
          <div
            class="mr-1 flex self-center"
          >
            <div
              class="mx-2 scale-110 rounded-md bg-highlight-green p-1 hover:cursor-pointer hover:scale-125 hover:bg-gray-700 hover:fill-highlight-green transition-all"
            >
              <div>
                <svg
                  height="20"
                  viewBox="0 96 960 960"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M395 771 226 601l50-50 119 118 289-288 50 51-339 339Z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="mx-2 scale-110 rounded-md bg-gray-200 p-1 hover:cursor-pointer hover:scale-125 hover:bg-highlight-red hover:fill-white transition-all"
            >
              <div>
                <svg
                  height="20"
                  viewBox="0 96 960 960"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m291 816-51-51 189-189-189-189 51-51 189 189 189-189 51 51-189 189 189 189-51 51-189-189-189 189Z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="py-2 flex justify-between border-b"
        >
          <div
            class="flex"
          >
             
            <a
              href="/profile/user-id-2"
            >
              <img
                alt="avatar"
                class="inline-block aspect-square rounded-[50%] self-center mt-1"
                src="image"
                style="width: 2rem;"
              />
            </a>
            <div
              class="mx-2"
            >
              <p>
                test user 2
              </p>
              <p
                class="text-xs text-gray-400"
              >
                0 seconds ago
              </p>
            </div>
          </div>
          <div
            class="mr-1 flex self-center"
          >
            <div
              class="mx-2 scale-110 rounded-md bg-highlight-green p-1 hover:cursor-pointer hover:scale-125 hover:bg-gray-700 hover:fill-highlight-green transition-all"
            >
              <div>
                <svg
                  height="20"
                  viewBox="0 96 960 960"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M395 771 226 601l50-50 119 118 289-288 50 51-339 339Z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="mx-2 scale-110 rounded-md bg-gray-200 p-1 hover:cursor-pointer hover:scale-125 hover:bg-highlight-red hover:fill-white transition-all"
            >
              <div>
                <svg
                  height="20"
                  viewBox="0 96 960 960"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m291 816-51-51 189-189-189-189 51-51 189 189 189-189 51 51-189 189 189 189-51 51-189-189-189 189Z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Contributor Request Menu: Snapshot > render menu open with no requests 1`] = `
<body>
  <div>
    <div
      class="absolute right-0 md:right-5 top-12 w-full max-h-96 md:w-96 rounded-md bg-white p-3 shadow border"
    >
      <div
        class="scrollbar-hide max-h-80 overflow-auto overflow-x-hidden"
      >
        <div
          class="sticky top-0 bg-white"
        >
          <p
            class="text-lg mb-2"
          >
            Contributor Requests
          </p>
          <hr />
        </div>
        <div
          class="flex w-full justify-center"
        >
          <p
            class="my-5 text-sm italic text-gray-400"
          >
            No contributor requests
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
